src/vm.rs

```diff
@@ -1488,6 +1496,56 @@ impl<'a, T: GuestStagePagingMode> FinalizedVm<'a, T> {
         Ok(0)
     }
 
+    fn handle_salus_test(
+        &self,
+        test_func: SalusTestFunction,
+        _active_pages: &ActiveVmPages<T>,
+    ) -> EcallResult<u64> {
+        use SalusTestFunction::*;
+        match test_func {
+            MemCopy(args) => self.guest_test_memcopy(args.to, args.from, args.len),
+        }
+    }
+
+    fn map_guest_range_in_umode_slot(
+        &self,
+        slot: u64,
+        addr: u64,
+        len: u64,
+        writable: bool,
+    ) -> EcallResult<(u64, UmodeMapping)> {
+        let base = PageSize::Size4k.round_down(addr);
+        let end = addr
+            .checked_add(len)
+            .ok_or(EcallError::Sbi(SbiError::InvalidParam))?;
+        let pages = self
+            .vm_pages()
+            .pin_shared_pages(
+                self.guest_addr_from_raw(base)?,
+                PageSize::num_4k_pages(end - base),
+            )
+            .map_err(EcallError::from)?;
+        let mapping = UmodeMapping::new(slot, pages, writable)
+            .map_err(|_| EcallError::Sbi(SbiError::Failed))?;
+        let vaddr = mapping.vaddr().bits() + (addr - base);
+        Ok((vaddr, mapping))
+    }
+
+    fn guest_test_memcopy(&self, to: u64, from: u64, len: u64) -> EcallResult<u64> {
+        let overlap = core::cmp::max(to, from) <= core::cmp::min(to + len - 1, from + len - 1);
+        if overlap {
+            return Err(EcallError::Sbi(SbiError::InvalidParam));
+        }
+        let (from_vaddr, _from_mapping) =
+            self.map_guest_range_in_umode_slot(0, from, len, false)?;
+        let (to_vaddr, _to_mapping) = self.map_guest_range_in_umode_slot(1, to, len, true)?;
+        // Safe because we checked that the two ranges do not overlap, and we mapped them for at
+        // least `len` bytes.
+        let request = unsafe { u_mode_api::UmodeRequest::memcopy(to_vaddr, from_vaddr, len) };
```

Comment:
So the issue I've been wondering during the development of this patch was: should we mark unsafe function that might make __umode__ unsafe?

The unsafety of memcopy here (and all the function that follows) is to guarantee the safety of umode. As umode will borrow mutable the slice starting at `to_addr`, in order to guarantee that the rust memory model in umode is respected we have to check that the two addresses do not overlap.

Reply:
Regarding the safety of the u64, this is a behaviour I have seen it in other calls that map guest addresses either in input or output, and they seem to be all marked unsafe.

Reply:
Safety is only relative to Salus' memory. So, in this case memcopy's safety comment is confusing because it's talking about not aliasing u-mode's memory with itself.

From the Salus side, this call is safe no matter what is passed as an address because of the guarantees made when mapping memory to u-mode (that there are no aliases created to memory salus can access).